### ISSUE TITLE
Investigate failure of URL parsing tests

### ISSUE THREAD

**Issue by gterzian** [servo/servo#26287]
Follow up from https://github.com/servo/servo/pull/26231

The PR marks some tests as FAIL, see https://github.com/servo/servo/pull/26231/files#diff-6c940171cdc241cde1caf796556840caR132

Actually that file contained a bunch of existing test failures already, related to things other than the use of `location`, and those could be investigated as well. 

For discussion, see https://github.com/servo/servo/pull/26231#issuecomment-618413191

**Comment by gterzian** [comment]
The spec is mainly at https://html.spec.whatwg.org/multipage/#parse-a-url

Which is used in various places related to the `/url/failure.html` test, for example via https://html.spec.whatwg.org/multipage/#the-location-interface:parse-a-url or https://html.spec.whatwg.org/multipage/#apis-for-creating-and-navigating-browsing-contexts-by-name:parse-a-url

**Comment by gterzian** [comment]
Also, the test is called `failure.html`, so we're not investigating why it fails, we're investigating why it fails to fail!

**Comment by gterzian** [comment]
Code would be in various places, such as: 

- https://github.com/servo/servo/blob/master/components/script/dom/location.rs

- https://github.com/servo/servo/blob/412bc1cddd8027596950164f3b396198bdcba25e/components/script/dom/windowproxy.rs#L455 (that one actually has a relevant TODO)

- other places where the equivalent of `url.join` is called.

**Comment by utsavoza** [comment]
@gterzian I see that the return type of [window.open()](https://github.com/servo/servo/blob/master/components/script/dom/window.rs#L659) is currently `Option<DomRoot<WindowProxy>>` as per the [spec](https://html.spec.whatwg.org/multipage/window-object.html#the-window-object) and [Window.webidl](https://github.com/servo/servo/blob/master/components/script/dom/webidls/Window.webidl#L43).

However, in the step 13 of the [window open steps](https://html.spec.whatwg.org/multipage/window-object.html#window-open-steps) we are expected to throw a "SyntaxError" DomException when the parse URL algorithm fails. So if the operation is fallible, shouldn't the IDL specify it as such?
```
[Throws] WindowProxy? open(optional USVString url = "", optional DOMString target = "_blank", optional [LegacyNullToEmptyString] DOMString features = "");
```
 As far as I can tell, in [Gecko we do consider the operation as fallible](https://github.com/mozilla/gecko-dev/blob/master/dom/webidl/Window.webidl#L241).

**Comment by gterzian** [comment]
Yes I think that's correct, good catch, so that would then indeed be a change required to implement the URL parsing changes.

**Comment by utsavoza** [comment]
So, I did update the window.open() method to return a fallible result and it seems we are able to pass quite a few url failure tests for window.open(). I am not sure if I should raise a PR that partially resolves the issue.

Also, I think the urls for which we had to mark some tests as FAIL for the location interface also got parsed correctly [here](https://github.com/servo/servo/blob/master/components/script/dom/windowproxy.rs#L453) for window.open(), though I haven't gone through the `url.join` implementation exhaustively.

**Comment by gterzian** [comment]
> I am not sure if I should raise a PR that partially resolves the issue.

yes I think a PR just for the window.open part would be great!

Let me know what you find out about the other tests...

**Comment by utsavoza** [comment]
@gterzian We'll need to re-open the issue as PR #26410 only partially resolves the issue.

**Comment by gterzian** [comment]
Ok, thanks! I didn't see that PR before it had already merged! :)

**Comment by utsavoza** [comment]
After digging a little bit, I was able narrowed down the issue to two things:
1. Currently, servo is pulling in `url` crate version 2.1.0 instead of 2.1.1 and based on [this PR](https://github.com/servo/rust-url/pull/537) itself (and updating the dependency it locally to verify), a lot of currently failing tests are resolved.
2. Based on the [spec](https://html.spec.whatwg.org/multipage/urls-and-fetching.html#parse-a-url), I think our implementation of some location setters (specifically `href`) incorrectly fetches the document URL instead of base URL for the entry settings object.
https://github.com/servo/servo/blob/6aec2c8594b15245b1e66b49e4c0e87176d3aad7/components/script/dom/location.rs#L254-L256

    And the document's implementation of `base_url` (specifically `fallback_base_url`) is currently incomplete:
https://github.com/servo/servo/blob/6aec2c8594b15245b1e66b49e4c0e87176d3aad7/components/script/dom/document.rs#L573-L579


@gterzian I wonder if I should address these in two separate PRs?

**Comment by jdm** [comment]
Addressing those separately makes sense to me. The fix for the fallback base url would also help in https://github.com/servo/servo/issues/24944#issuecomment-559878490 which has some additional information; it's going to require passing some information around about the browsing context creator when loading iframes so documents can store it appropriately.

**Comment by utsavoza** [comment]
Thanks, since I have tried updating the dependency locally, I can raise a PR for it right away (though I haven't run all the wpt tests), updating the `fallback_url` might need some more work on my end.

**Comment by utsavoza** [comment]
I'd like to reiterate the observations in https://github.com/servo/servo/pull/26231#issuecomment-617982016. As per the [README](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/url/README.md):

> In addition to testing that parsing input against base gives the result, a test harness for the URL constructor (or similar APIs) should additionally test the following pattern: if failure is true, parsing about:blank against input must give failure. This tests that the logic for converting base URLs into strings properly fails the whole parsing algorithm if the base URL cannot be parsed.

For all the tests that are expected to fail, the input must be parsed against `about:blank` as base url. However, during the test execution, the base url that get's picked up is `http://web-platform.test:8000/url/failure.html`. Even though this eliminates several bad urls, there are still [some edge cases](https://github.com/servo/servo/pull/26231/files#diff-6c940171cdc241cde1caf796556840caR132) that gets parsed correctly (according to the [spec](https://url.spec.whatwg.org/#concept-basic-url-parser)) against a valid base url but fails against `about:blank`.

**Comment by mrobinson** [comment]
I don't see these expectations any longer so we can close this issue.
